File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/warnings/Unused.scala
package org.enso.compiler.core.ir
package expression
package warnings

/** Warnings about unused language entities. */
sealed trait Unused extends Warning {
  val name: Name
}

object Unused {

  /** A warning about an unused function argument.
    *
    * @param name the name that is unused
    */
  sealed case class FunctionArgument(override val name: Name) extends Unused {
    override def message: String = s"Unused function argument ${name.name}."

    override def diagnosticKeys(): Array[Any] = Array(name.name)

    override def toString: String = s"Unused.FunctionArgument(${name.name})"

    override val location: Option[IdentifiedLocation] = name.location
  }

  sealed case class PatternBinding(override val name: Name) extends Unused {
    override def message: String = s"Unused pattern binding ${name.name}."

    override def diagnosticKeys(): Array[Any] = Array(name.name)

    override def toString: String = s"Unused.PatternBinding(${name.name})"

    override val location: Option[IdentifiedLocation] = name.location
  }

  /** A warning about an unused binding.
    *
    * @param name the name that is unused
    */
  sealed case class Binding(override val name: Name) extends Unused {
    override def message: String = s"Unused variable ${name.name}."

    override def diagnosticKeys(): Array[Any] = Array(name.name)

    override def toString: String = s"Unused.Binding(${name.name})"

    override val location: Option[IdentifiedLocation] = name.location
  }
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/expression/warnings/Unused.java
package org.enso.compiler.core.ir.expression.warnings;

import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.Name;
import org.enso.compiler.core.ir.Warning;
import scala.Function1;

/** Warnings about unused language entities. */
public sealed interface Unused extends Warning
    permits Unused.FunctionArgument, Unused.PatternBinding, Unused.Binding {

  /** The name that is unused. */
  Name name();

  /**
   * A warning about an unused function argument.
   *
   * @param name the name that is unused
   */
  record FunctionArgument(Name name) implements Unused {
    @Override
    public String message(Function1<IdentifiedLocation, String> source) {
      return "Unused function argument " + name.name() + ".";
    }

    @Override
    public Object[] diagnosticKeys() {
      return new Object[] {name.name()};
    }

    @Override
    public IdentifiedLocation identifiedLocation() {
      return name.identifiedLocation();
    }

    @Override
    public String toString() {
      return "Unused.FunctionArgument(" + name.name() + ")";
    }
  }

  /**
   * A warning about an unused pattern binding.
   *
   * @param name the name that is unused
   */
  record PatternBinding(Name name) implements Unused {
    @Override
    public String message(Function1<IdentifiedLocation, String> source) {
      return "Unused pattern binding " + name.name() + ".";
    }

    @Override
    public Object[] diagnosticKeys() {
      return new Object[] {name.name()};
    }

    @Override
    public IdentifiedLocation identifiedLocation() {
      return name.identifiedLocation();
    }

    @Override
    public String toString() {
      return "Unused.PatternBinding(" + name.name() + ")";
    }
  }

  /**
   * A warning about an unused binding.
   *
   * @param name the name that is unused
   */
  record Binding(Name name) implements Unused {
    @Override
    public String message(Function1<IdentifiedLocation, String> source) {
      return "Unused variable " + name.name() + ".";
    }

    @Override
    public Object[] diagnosticKeys() {
      return new Object[] {name.name()};
    }

    @Override
    public IdentifiedLocation identifiedLocation() {
      return name.identifiedLocation();
    }

    @Override
    public String toString() {
      return "Unused.Binding(" + name.name() + ")";
    }
  }
}
